### PR TITLE
Support for interval based Checkpointing

### PR DESCRIPTION
**Pull request by basavaraj29**
Introduced support for checkpointing periodically, checkpoint disabled by default. 

when enabled with `training.checkpoint.interval` set to `k`, 
writes the checkpointed model into the `<model_dir>/checkpoint_x` directory, where x=n*k. tested resuming training from the checkpointed model, works fine. 

previous checkpoints can be preserved by setting `training.checkpoint.save_prev_num`. the specified number of previous checkpoints would be preserved. 

Checkpointing Overhead
for freebase86m dataset, on cosmos 1 machine, the epoch time is around 37.2 mins (2234 secs) and the checkpointing time is around 1 min. ~3% overhead if we checkpoint for each epoch. 
```
[05/21/22 06:42:12.880] Epoch Runtime: 2233951ms
[05/21/22 06:43:12.637] Checkpoint Time: 59756ms
```

### CODE REVIEW

File: src/cpp/include/configuration/config.h [hunk 1]
```diff
@@ -152,6 +152,7 @@ struct TrainingConfig {
     shared_ptr<CheckpointConfig> checkpoint = nullptr;
```
Comment (JasonMoho) [inline src/cpp/include/configuration/config.h:152]: Is this checkpoint config used anywhere?

My original purpose was to have this as nested option for the checkpointing stuff. E.g.

```
training:
  ...
  checkpoint:
    interval: 10
```

So either this should be deleted or the other checkpoint options should be merged into it. What do you think?

File: src/python/tools/configuration/marius_config.py [hunk 2]
```diff
@@ -886,7 +887,7 @@ def load_config(input_config_path, save=False):
     
     if output_config.training.resume_from_checkpoint != "" and output_config.training.resume_from_checkpoint[-1] != "/":
         output_config.training.resume_from_checkpoint += "/"
-
+
```
Comment (JasonMoho) [inline src/python/tools/configuration/marius_config.py:890]: Nit: Some random whitespace changes in this file

File: src/cpp/src/marius.cpp [hunk 3]
```diff
@@ -161,9 +160,14 @@ void marius_train(shared_ptr<MariusConfig> marius_config) {
                 evaluator->evaluate(false);
             }
         }
-    }
 
-    metadata.num_epochs = dataloader->epochs_processed_;
+
+        metadata.num_epochs = dataloader->epochs_processed_;
+        if(marius_config->training->checkpoint_after_epochs > 0 && (epoch + 1) % marius_config->training->checkpoint_after_epochs == 0 &&
```
Comment (JasonMoho) [inline src/cpp/src/marius.cpp:166]: nit: there should be a space after the `if`

Also add a new line after the first `&&` so that each condition is on a new line. Makes it a bit more readable.

File: src/cpp/src/storage/checkpointer.cpp [hunk 4]
```diff
@@ -15,6 +16,54 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
...
+    std::filesystem::create_directory(tmp_checkpoint_dir);
+    
+    std::string new_embeddings_file = tmp_checkpoint_dir
+                                        + PathConstants::embeddings_file 
+                                        + PathConstants::file_ext;
+    std::string new_embeddings_state_file = tmp_checkpoint_dir
+                                            + PathConstants::embeddings_state_file
+                                            + PathConstants::file_ext;
+
+    if(std::filesystem::exists(prev_checkpoint_dir)) {
```
Comment (JasonMoho) [inline src/cpp/src/storage/checkpointer.cpp:33]: Why is this code needed? The embeddings/state from the previous checkpoint are moved to the new checkpoint before being overwritten by `this-save()`? 

Am I missing something or should this whole if/else be deleted.

File: src/cpp/src/storage/checkpointer.cpp [hunk 5]
```diff
@@ -15,6 +16,54 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
...
+            std::filesystem::copy(embeddings_state_file, new_embeddings_state_file);
+        }
+    }
+
+    this->save(tmp_checkpoint_dir, checkpoint_meta);
+
+    string final_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "/";
+    std::filesystem::rename(tmp_checkpoint_dir, final_checkpoint_dir);
+    
+    std::filesystem::remove_all(prev_checkpoint_dir);
```
Comment (JasonMoho) [inline src/cpp/src/storage/checkpointer.cpp:64]: Not sure it makes sense to delete the previous checkpoint always? A user might want to keep their previous checkpoints around. 

Maybe we should have an option for this.

**Review by JasonMoho — CHANGES_REQUESTED** [review]
Make sure to check that your `if` statements have proper whitespace.

We should think what to do with the training.checkpoint options. We may want to use them to give users control whether they want to:
- keep the optimizer state around
- only keep the latest checkpoint

File: src/cpp/src/storage/checkpointer.cpp (hunk 4 above)
Comment (basavaraj29) [reply src/cpp/src/storage/checkpointer.cpp]: `this->save()` doesn't save the params to the checkpoint dir, it saves them to the `model_dir` i guess.

In this function `Checkpointer::save(...)`
`model->save()` function takes in a path argument and saves the model to the specified directory, while `storage_->storage_ptrs_.node_embeddings->write()` doesn't.

File: src/cpp/src/storage/checkpointer.cpp [hunk 6]
```diff
@@ -33,88 +61,87 @@ void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
...
-//                if (fileExists(node_state_filename)) {
-//                    copyFile(node_state_filename, output_node_state_filename);
-//                }
-//            }
-//        }
-//        copyFile(storage_->base_directory_ + PathConstants::config_file, checkpoint_dir + PathConstants::config_file);
-//
-//        copyFile(storage_->base_directory_ + PathConstants::checkpoint_metadata_file, checkpoint_dir + PathConstants::checkpoint_metadata_file);
-//    }
+    //    if (storage_->base_directory_ != checkpoint_dir) {
```
Comment (basavaraj29) [inline src/cpp/src/storage/checkpointer.cpp:64]: auto-formatter changes. Also, should we remove this commented code?

File: src/cpp/include/configuration/config.h [hunk 7]
```diff
@@ -113,6 +113,7 @@ struct CheckpointConfig {
     bool save_edges;
```
Comment (JasonMoho) [inline src/cpp/include/configuration/config.h:113]: Could you delete the unused configuration options

File: src/cpp/src/storage/checkpointer.cpp (hunk 6 above)
Comment (JasonMoho) [reply src/cpp/src/storage/checkpointer.cpp:64]: Yep this code is safe to delete

File: src/cpp/src/storage/checkpointer.cpp [hunk 8]
```diff
@@ -16,58 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
...
     }
 
     this->save(tmp_checkpoint_dir, checkpoint_meta);
 
     string final_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "/";
     std::filesystem::rename(tmp_checkpoint_dir, final_checkpoint_dir);
-    
-    std::filesystem::remove_all(prev_checkpoint_dir);
+
+    int delete_checkpoint = epochs - this->config_->save_prev_num * this->config_->interval;
```
Comment (JasonMoho) [inline src/cpp/src/storage/checkpointer.cpp:41]: I think for simplicity his save_prev_num option can be deleted and we just keep around all checkpoints for now. 

In a future PR it would be nice to have a feature called "save_best" which will save the checkpoint which performs best on the valid/test set.

File: src/cpp/src/marius.cpp [hunk 9]
```diff
@@ -18,9 +18,7 @@ void encode_and_export(shared_ptr<DataLoader> dataloader, shared_ptr<Model> mode
         graph_encoder = std::make_shared<PipelineGraphEncoder>(dataloader, model, marius_config->evaluation->pipeline);
     }
 
-    string filename = marius_config->storage->model_dir
-                      + PathConstants::encoded_nodes_file
-                      + PathConstants::file_ext;
+    string filename = marius_config->storage->model_dir + PathConstants::encoded_nodes_file + PathConstants::file_ext;
```
Comment (JasonMoho) [inline src/cpp/src/marius.cpp:21]: revert

File: src/cpp/src/marius.cpp [hunk 10]
```diff
@@ -109,14 +106,12 @@ std::tuple<shared_ptr<Model>, shared_ptr<GraphModelStorage>, shared_ptr<DataLoad
     initialization_timer.stop();
     int64_t initialization_time = initialization_timer.getDuration();
 
-    SPDLOG_INFO("Initialization Complete: {}s", (double) initialization_time / 1000);
+    SPDLOG_INFO("Initialization Complete: {}s", (double)initialization_time / 1000);
```
Comment (JasonMoho) [inline src/cpp/src/marius.cpp:109]: revert

File: src/cpp/src/marius.cpp [hunk 11]
```diff
@@ -198,8 +196,7 @@ void marius_eval(shared_ptr<MariusConfig> marius_config) {
 }
 
 void marius(int argc, char *argv[]) {
-
-    (void) argc;
+    (void)argc;
```
Comment (JasonMoho) [inline src/cpp/src/marius.cpp:199]: revert

File: src/cpp/src/storage/checkpointer.cpp [hunk 12]
```diff
@@ -15,10 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
...
 
-    if (checkpoint_meta.has_model) {
+    std::string new_embeddings_file = tmp_checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string new_embeddings_state_file = tmp_checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    std::string embeddings_file = checkpoint_dir + PathConstants::embeddings_file + PathConstants::file_ext;
+    std::string embeddings_state_file = checkpoint_dir + PathConstants::embeddings_state_file + PathConstants::file_ext;
+
+    if (std::filesystem::exists(embeddings_file)) {
+        if (this->config_->save_encoded) std::filesystem::copy(embeddings_file, new_embeddings_file);
```
Comment (JasonMoho) [inline src/cpp/src/storage/checkpointer.cpp:32]: The save_encoded option can be removed and the embeddings can be copied always.

The original idea behind the save_encoded option was to save the output of the last layer of the encoder at each checkpoint. But it's fine to not include that functionality.

File: src/cpp/src/storage/checkpointer.cpp [hunk 13]
```diff
@@ -15,10 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
     config_ = config;
 }
 
-void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
+void Checkpointer::create_checkpoint(string checkpoint_dir, CheckpointMeta checkpoint_meta, int epochs) {
+    string tmp_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "_tmp/";
```
Comment (JasonMoho) [inline src/cpp/src/storage/checkpointer.cpp:22]: Why create a tmp directory and not just put the files into the checkpoint directory directly?

File: src/cpp/src/storage/checkpointer.cpp [hunk 14]
```diff
@@ -15,10 +18,35 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
...
+
+    int delete_checkpoint = epochs - this->config_->save_prev_num * this->config_->interval;
+    if (delete_checkpoint > 0) {
+        std::string delete_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(delete_checkpoint) + "/";
+        std::filesystem::remove_all(delete_checkpoint_dir);
+    }
+}
+
+void Checkpointer::save(string checkpoint_dir, CheckpointMeta checkpoint_meta) {
+    if (checkpoint_meta.has_model) {
```
Comment (JasonMoho) [inline src/cpp/src/storage/checkpointer.cpp:49]: The CheckpointMeta struct and corresponding file could be cleaned up. Can you 1) delete the unused options in CheckpointMeta and 2) have the metadata file include the keys so it is more interpretable. 

I believe the current format just has the values and no keys. So the new version should be something like

```
has_state: 1
link_prediction: 1
...
```

However adding the keys to the checkpoint metadata file may make the parsing code significantly more complicated (C++ isn't great at this stuff), so if this is the case you can skip that part and just delete the unused options.

**Review by JasonMoho — CHANGES_REQUESTED** [review]
On top of the review comments can you add a test for this functionality?

File: src/cpp/src/storage/checkpointer.cpp (hunk 13 above)
Comment (basavaraj29) [reply src/cpp/src/storage/checkpointer.cpp:22]: a bit far-fetched. we could crash in the middle of the process. if so, we wouldn't want to use the half-written checkpoint. hence went with the rename option which I believe is atomic. so if there exists a directory of the form `checkpoint_x`, we can be sure that the checkpoint is valid.

**Review by JasonMoho — APPROVED** [review]
